threatingestor/operators/misp.py

```diff
@@ -32,69 +37,102 @@ def __init__(self, url, key, ssl=True, tags=None, artifact_types=None, filter_st
+            event = self.handle_domain(event, artifact)
         if isinstance(artifact, threatingestor.artifacts.Hash):
-            self.handle_hash(artifact)
+            event = self.handle_hash(event, artifact)
         elif isinstance(artifact, threatingestor.artifacts.IPAddress):
-            self.handle_ipaddress(artifact)
+            event = self.handle_ipaddress(event, artifact)
         if isinstance(artifact, threatingestor.artifacts.URL):
-            self.handle_url(artifact)
+            event = self.handle_url(event, artifact)
         elif isinstance(artifact, threatingestor.artifacts.YARASignature):
-            self.handle_yarasignature(artifact)
+            event = self.handle_yarasignature(event, artifact)
+
+        self._update_or_create_event(event)
 
+    def _update_or_create_event(self, event):
+        """Update or create an event for the artifact."""
+        event_dict = event.to_dict()
+        attributes = event_dict.get("Attribute", [])
+        if not attributes:
+            return
+
+        if event_dict.get("uuid") is None:
+            self.api.add_event(event)
+        else:
+            self.api.update_event(event)
+
+    def _find_or_create_event(self, artifact):
+        """Find or create an event for the artifact."""
+        event = self._find_event(artifact)
+        if event is not None:
+            return event
+
+        return self._create_event(artifact)
+
+    def _find_event(self, artifact):
+        """Find an event which has the same refetrence link, return an Event object."""
+        events = self.api.search(
+            "events",
+            limit=1,
+            type_attribute="link",
+            value=artifact.reference_link,
+            pythonify=True
+        )
+        if len(events) == 1:
+            return events[0]
+
+        return None
 
     def _create_event(self, artifact):
         """Create an event in MISP, return an Event object."""
-        event = self.api.new_event(info=self.event_info.format(
-            source_name=artifact.source_name))
+        event = pymisp.MISPEvent()
+        event.info = self.event_info.format(source_name=artifact.source_name)
 
         # Add tags.
         for tag in self.tags:
-            self.api.add_tag(event, tag)
+            event.add_tag(tag)
 
         # Add references.
-        self.api.add_internal_link(event, artifact.reference_link)
-        self.api.add_internal_text(event, artifact.reference_text)
-        self.api.add_internal_other(event, f'source:{artifact.source_name}')
+        if artifact.reference_link != "":
+            event.add_attribute("link", artifact.reference_link)
+        if artifact.reference_text != "":
+            event.add_attribute("text", artifact.reference_text)
+        if artifact.source_name != "":
+            event.add_attribute("other",  f'source:{artifact.source_name}')
 
         return event
 
-
-    def handle_domain(self, domain):
+    def handle_domain(self, event, domain):
```

Comment:
When adding attributes to a function or method, it's a better idea to add them to the end and make them keyword arguments, just in case there's a call to `handle_domain()` somewhere that doesn't get updated. It looks like you handled all the cases in operators, but there could be ThreatIngestor user code that could break.

Reply:
The same goes for `handle_hash()`, `handle_ipaddress()`, `handle_url()`, and `handle_yarasignature()`.